// common-services/dist/log.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_LOG_ROUTING = exports.DEFAULT_LOGGER = exports.StdStream = void 0;

var _knifecycle = require("knifecycle");

function noop() {
  return undefined;
}

let StdStream;
exports.StdStream = StdStream;

(function (StdStream) {
  StdStream["STD_OUT"] = "STD_OUT";
  StdStream["STD_ERR"] = "STD_ERR";
})(StdStream || (exports.StdStream = StdStream = {}));

// Fallbacking to silent logs making logging
// opt-in. It appears to be the best default
// since logs are easy to get but won't disturb
// tests.
const DEFAULT_LOGGER = {
  debug: noop,
  // eslint-disable-line
  error: noop,
  // eslint-disable-line
  info: noop,
  // eslint-disable-line
  warning: noop // eslint-disable-line

};
exports.DEFAULT_LOGGER = DEFAULT_LOGGER;
const DEFAULT_LOG_ROUTING = {
  error: StdStream.STD_ERR,
  // The stack type allows to filter logs in testing
  // since the stack files paths vary between systems
  // and it is annoying to filter them
  stack: StdStream.STD_ERR,
  warning: StdStream.STD_ERR,
  info: StdStream.STD_OUT
};
/* Architecture Note #1.1: Logging

I prefer using a unique function with the log type
 in parameter instead of several methods for each
 log types. It is far easyer to mock and to assert
 on logs in my tests.

If provided, I route debug messages to the `debug`
 node module.

*/

exports.DEFAULT_LOG_ROUTING = DEFAULT_LOG_ROUTING;

var _default = (0, _knifecycle.singleton)((0, _knifecycle.service)(initLog, "log", ["?LOG_ROUTING", "?logger", "?debug"]));
/**
 * Instantiate the logging service
 * @name initLog
 * @function
 * @param  {Object}   services
 * The services to inject
 * @param  {Object}   [services.logger = DEFAULT_LOGGER]
 * The logger to use
 * @param  {Function} [services.debug = noop]
 * A debugging function
 * @return {Promise<Function>}
 * A promise of the logging function
 * @example
 * import initLog from 'common-services/dist/log';
 *
 * const log = await initLog({
 *   logger: require('winston'),
 *   debug: require('debug')('myapp'),
 *  });
 */


exports.default = _default;

async function initLog({
  LOG_ROUTING = DEFAULT_LOG_ROUTING,
  logger = DEFAULT_LOGGER,
  debug = noop
}) {
  log('debug', '👣 - Logging service initialized.');
  return log;
  /**
   * Logging function
   * @param  {String}  type
   * Log type
   * @param  {...*}    args
   * Log contents
   * @return {void}
   * @example
   * log('debug', 'Luke, I am your father!')
   */

  function log(type, ...args) {
    if (LOG_ROUTING[type] === StdStream.STD_ERR) {
      logger.error(...args);
      return;
    }

    if (LOG_ROUTING[type] === StdStream.STD_OUT) {
      logger.info(...args);
      return;
    }

    debug(...args);
  }
}
